web-server/src/content/Service/SystemLogs.tsx

```diff
@@ -16,82 +16,171 @@ import { parseLogLine } from '@/utils/logFormatter';
+    }, 300);
+    return () => clearTimeout(timer);
+  }, [searchQuery]);
 
+  const scrollToBottom = useCallback(() => {
     if (containerRef.current) {
-      observer.observe(containerElement);
+      containerRef.current.scrollTo({
+        top: containerRef.current.scrollHeight,
+        behavior: 'smooth'
+      });
+    }
+  }, []);
+
+  const updateHighlight = useCallback((element: HTMLElement | null) => {
+    if (currentHighlightRef.current) {
+      currentHighlightRef.current.style.backgroundColor = 'yellow';
+      currentHighlightRef.current.style.color = 'black';
+    }
+
+    if (element) {
+      element.style.backgroundColor = 'rgb(255, 161, 10)';
+      element.style.color = 'white';
+      currentHighlightRef.current = element;
+    } else {
+      currentHighlightRef.current = null;
     }
+  }, []);
 
-    return () => {
-      if (containerElement) {
-        observer.unobserve(containerElement);
+  const handleSearch = useCallback((query: string) => {
+    setSearchQuery(query);
+    isSearchingRef.current = !!query;
+  
+    if (!query || query.length < 3) {
+      setCurrentMatch(0);
+      currentMatchRef.current = 0;
+      setTotalMatches(0);
+      totalMatchesRef.current = 0;
+      setHighlightedElements([]);
+      highlightedElementsRef.current = [];
+      updateHighlight(null);
+      return;
+    }
+  }, [updateHighlight]);
+  
+  useLayoutEffect(() => {
+    if (!debouncedSearchQuery || debouncedSearchQuery.length < 3) return;
+  
+    const elements = Array.from(
+      containerRef.current?.querySelectorAll('span[style*="background-color: yellow"]') ?? []
+    ) as HTMLElement[];
+  
+    setHighlightedElements(elements);
+    highlightedElementsRef.current = elements;
+    const newTotalMatches = elements.length;
+    setTotalMatches(newTotalMatches);
+    totalMatchesRef.current = newTotalMatches;
+    
+    if (currentMatchRef.current === 0) {
+      setCurrentMatch(newTotalMatches ? 1 : 0);
+      currentMatchRef.current = newTotalMatches ? 1 : 0;
+    }
+  }, [debouncedSearchQuery]);
+
+  const handleNavigate = useCallback((direction: 'prev' | 'next') => {
+    if (highlightedElementsRef.current.length === 0 || isNavigatingRef.current) return;
+
+    isNavigatingRef.current = true;
+    requestAnimationFrame(() => {
+      let newIndex = currentMatchRef.current;
+      const total = totalMatchesRef.current;
+
+      if (direction === 'next') {
+        newIndex = newIndex < total ? newIndex + 1 : 1;
+      } else {
+        newIndex = newIndex > 1 ? newIndex - 1 : total;
       }
-    };
-  }, [showScrollDownButton]);
+
+      setCurrentMatch(newIndex);
+      currentMatchRef.current = newIndex;
+      const element = highlightedElementsRef.current[newIndex - 1];
+      if (element) {
+        element.scrollIntoView({ behavior: 'smooth', block: 'center' });
+        updateHighlight(element);
+      }
+      isNavigatingRef.current = false;
+    });
+  }, [updateHighlight]);
```

Comment:
This is an example of adding hacky code to make existing hacky code work.

Using a proper state machine architecture will make this SO MUCH EASIER to handle.